flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java

```diff
@@ -0,0 +1,415 @@
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.TaskExecutionStateTransition;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.jobmanager.PartitionProducerDisposedException;
+import org.apache.flink.runtime.jobmaster.SerializedInputSplit;
+import org.apache.flink.runtime.messages.FlinkJobNotFoundException;
+import org.apache.flink.runtime.messages.checkpoint.DeclineCheckpoint;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequest;
+import org.apache.flink.runtime.operators.coordination.CoordinationResponse;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.query.KvStateLocation;
+import org.apache.flink.runtime.query.UnknownKvStateLocation;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.KvStateHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.util.FlinkException;
+
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+
+/**
+ * Abstract state class which contains an {@link ExecutionGraph} and the required handlers to
+ * execute common operations.
+ */
+abstract class StateWithExecutionGraph implements State {
+    private final Context context;
+
+    private final ExecutionGraph executionGraph;
+
+    private final ExecutionGraphHandler executionGraphHandler;
+
+    private final OperatorCoordinatorHandler operatorCoordinatorHandler;
+
+    private final KvStateHandler kvStateHandler;
+
+    private final Logger logger;
+
+    StateWithExecutionGraph(
+            Context context,
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger) {
+        this.context = context;
+        this.executionGraph = executionGraph;
+        this.executionGraphHandler = executionGraphHandler;
+        this.operatorCoordinatorHandler = operatorCoordinatorHandler;
+        this.kvStateHandler = new KvStateHandler(executionGraph);
+        this.logger = logger;
+
+        FutureUtils.assertNoException(
+                executionGraph
+                        .getTerminationFuture()
+                        .thenAcceptAsync(
+                                jobStatus -> {
+                                    operatorCoordinatorHandler.disposeAllOperatorCoordinators();
```

Comment:
I removed the `onLeave` method because of this discussion: https://github.com/tillrohrmann/flink/pull/24/files/4cb7aa868ab904d8610b2141e5998b8c61af02b0#r568477590

I added the onLeave method again, as it is the cleanest approach to handling the component lifecycle. I also added a test for it.

Reply:
Could we ease the problems we had in the past by passing the state _class_ instead of the state itself?

Reply:
Hmm, this wouldn't solve a lot of problems other than not giving direct access to a `State` instance. What else would it help us with?

Reply:
We don't have to construct that instance, nor any of its dependencies like an execution graph.
IOW, it behaves like passing in an enum describing which state we are in, without us having to create another enum.

Reply:
Actually I'd say it solves all the issues we had.

Reply:
We then could also call `onLeave` without having to construct the new state beforehand in the scheduler.
This would outright eliminate the possibility of a state constructor having some side effect that is executed before we have fully left another state. `onEnter` would then no longer be necessary because states would be free to do anything in the constructor.

Reply:
I think you've convinced me. Then let's do it like this.

Reply:
I'll update the PR!

Reply:
I have not removed the onEnter method, as this would change the behavior of the current DeclarativeScheduler.transitionToState() (onLeave gets called before onEnter, and the transition (hence the call to onEnter happens only to states different than the current)

Reply:
This can be refactored as a follow up.